File: products/compliance/src/analyzers/secret-scanner.ts
import type { Analyzer, Finding, RuleDefinition, ScannedFile } from '../types.js';
import { readFileSync } from 'node:fs';

/**
 * Pattern definition linking a regex to a rule pattern identifier.
 */
interface SecretPattern {
  name: string;
  regex: RegExp;
  rulePattern: string;
}

const TEXT_EXTENSIONS = new Set([
  'ts', 'js', 'py', 'go', 'java', 'rb',
  'yaml', 'yml', 'json', 'toml',
  'env', 'cfg', 'conf', 'ini', 'xml', 'properties',
]);

const MAX_FILE_SIZE = 500 * 1024; // 500 KB

const SECRET_PATTERNS: SecretPattern[] = [
  // AWS Access Key
  {
    name: 'AWS Access Key',
    regex: /AKIA[0-9A-Z]{16}/g,
    rulePattern: 'hardcoded-credential',
  },
  // AWS Secret Key (high-entropy string following common assignment patterns)
  {
    name: 'AWS Secret Key',
    regex: /(?:aws_secret|AWS_SECRET|secret_key|SECRET_KEY)\s*[=:]\s*['"]?([A-Za-z0-9/+=]{40})['"]?/g,
    rulePattern: 'hardcoded-credential',
  },
  // GitHub Token
  {
    name: 'GitHub Token',
    regex: /gh[ps]_[A-Za-z0-9_]{36,}/g,
    rulePattern: 'api-token',
  },
  // Private Key
  {
    name: 'Private Key',
    regex: /-----BEGIN\s+(?:RSA|EC|DSA|PGP)?\s*PRIVATE KEY-----/g,
    rulePattern: 'private-key',
  },
  // JWT Token
  {
    name: 'JWT Token',
    regex: /eyJ[A-Za-z0-9\-_]+\.eyJ[A-Za-z0-9\-_]+\.[A-Za-z0-9\-_.+/=]+/g,
    rulePattern: 'api-token',
  },
  // Slack Token
  {
    name: 'Slack Token',
    regex: /xox[bpras]-[0-9a-zA-Z-]+/g,
    rulePattern: 'api-token',
  },
  // Stripe Key
  {
    name: 'Stripe Key',
    regex: /sk_(?:live|test)_[0-9a-zA-Z]{24,}/g,
    rulePattern: 'api-token',
  },
  // Anthropic Key
  {
    name: 'Anthropic Key',
    regex: /sk-ant-[a-zA-Z0-9-]+/g,
    rulePattern: 'api-token',
  },
  // Generic Password (case insensitive)
  {
    name: 'Generic Password',
    regex: /(?:password|passwd|pwd|secret)\s*[=:]\s*['"][^'"]{4,}['"]/gi,
    rulePattern: 'hardcoded-credential',
  },
  // Generic API Key (case insensitive)
  {
    name: 'Generic API Key',
    regex: /(?:api[_-]?key|apikey)\s*[=:]\s*['"][^'"]{8,}['"]/gi,
    rulePattern: 'hardcoded-credential',
  },
  // Database URL with credentials
  {
    name: 'Database URL',
    regex: /(?:mongodb|postgres|mysql|redis):\/\/[^\s'"]+:[^\s'"]+@/g,
    rulePattern: 'hardcoded-credential',
  },
  // Google API Key
  {
    name: 'Google API Key',
    regex: /AIza[0-9A-Za-z\-_]{35}/g,
    rulePattern: 'api-token',
  },
  // Heroku API Key
  {
    name: 'Heroku API Key',
    regex: /[0-9a-fA-F]{8}-[0-9a-fA-F]{4}-[0-9a-fA-F]{4}-[0-9a-fA-F]{4}-[0-9a-fA-F]{12}/g,
    rulePattern: 'api-token',
  },
  // SendGrid API Key
  {
    name: 'SendGrid API Key',
    regex: /SG\.[0-9A-Za-z\-_]{22,}\.[0-9A-Za-z\-_]{22,}/g,
    rulePattern: 'api-token',
  },
  // Twilio API Key
  {
    name: 'Twilio API Key',
    regex: /SK[0-9a-fA-F]{32}/g,
    rulePattern: 'api-token',
  },
  // Mailgun API Key
  {
    name: 'Mailgun API Key',
    regex: /key-[0-9a-zA-Z]{32}/g,
    rulePattern: 'api-token',
  },
];

/**
 * Compute Shannon entropy of a string.
 */
function shannonEntropy(str: string): number {
  const freq = new Map<string, number>();
  for (const ch of str) freq.set(ch, (freq.get(ch) ?? 0) + 1);
  let entropy = 0;
  for (const count of freq.values()) {
    const p = count / str.length;
    entropy -= p * Math.log2(p);
  }
  return entropy;
}

/**
 * Redact the middle of a secret, keeping first and last 4 characters.
 */
function redact(secret: string): string {
  if (secret.length <= 8) return '****';
  return secret.slice(0, 4) + '****' + secret.slice(-4);
}

/**
 * Check if a string looks like it could be hex-encoded.
 */
function isHexLike(str: string): boolean {
  return /^[0-9a-fA-F]+$/.test(str);
}

/**
 * Check if a string looks like it could be base64-encoded.
 */
function isBase64Like(str: string): boolean {
  return /^[A-Za-z0-9+/=\-_]+$/.test(str);
}

/**
 * Check if a file extension is a text-like extension we should scan.
 */
function isTextExtension(ext: string): boolean {
  return TEXT_EXTENSIONS.has(ext.toLowerCase());
}

/**
 * Detect high-entropy strings that may be secrets.
 */
function detectHighEntropyStrings(line: string): string[] {
  const results: string[] = [];
  // Match quoted strings and bare assignments that look like secrets
  const candidates = line.matchAll(/['"]([A-Za-z0-9+/=\-_]{20,})['"]/g);
  for (const match of candidates) {
    const candidate = match[1];
    if (candidate.length < 20) continue;

    const threshold = isHexLike(candidate) ? 4.5 : isBase64Like(candidate) ? 5.0 : 5.0;
    const entropy = shannonEntropy(candidate);
    if (entropy > threshold) {
      results.push(candidate);
    }
  }
  return results;
}

interface MatchResult {
  patternName: string;
  rulePattern: string;
  matched: string;
  line: number;
  column: number;
}

export class SecretScanner implements Analyzer {
  name = 'secret-scanner';

  async analyze(files: ScannedFile[], rules: RuleDefinition[]): Promise<Finding[]> {
    // Filter rules relevant to this analyzer
    const myRules = rules.filter((r) => r.analyzer === this.name);
    if (myRules.length === 0) return [];

    // Build a lookup from pattern -> rules
    const rulesByPattern = new Map<string, RuleDefinition[]>();
    for (const rule of myRules) {
      const existing = rulesByPattern.get(rule.pattern) ?? [];
      existing.push(rule);
      rulesByPattern.set(rule.pattern, existing);
    }

    const findings: Finding[] = [];

    for (const file of files) {
      // Skip non-text files
      if (!isTextExtension(file.extension)) continue;
      // Skip large files
      if (file.size > MAX_FILE_SIZE) continue;

      let content: string;
      try {
        content = readFileSync(file.path, 'utf-8');
      } catch {
        // Skip files that can't be read (binary, permissions, etc.)
        continue;
      }

      // Quick binary check: if content contains null bytes, skip
      if (content.includes('\0')) continue;

      const lines = content.split('\n');
      const matches: MatchResult[] = [];

      for (let lineIdx = 0; lineIdx < lines.length; lineIdx++) {
        const line = lines[lineIdx];

        // Run each secret pattern against the line
        for (const pattern of SECRET_PATTERNS) {
          // Reset regex lastIndex since we reuse global regexes
          pattern.regex.lastIndex = 0;
          let match: RegExpExecArray | null;
          while ((match = pattern.regex.exec(line)) !== null) {
            matches.push({
              patternName: pattern.name,
              rulePattern: pattern.rulePattern,
              matched: match[0],
              line: lineIdx + 1,
              column: match.index + 1,
            });
          }
        }

        // Check for high-entropy strings
        const highEntropyMatches = detectHighEntropyStrings(line);
        for (const heMatch of highEntropyMatches) {
          // Avoid duplicate findings: skip if already matched by a specific pattern
          const alreadyMatched = matches.some(
            (m) => m.line === lineIdx + 1 && line.includes(m.matched) && heMatch.includes(m.matched),
          );
          if (!alreadyMatched) {
            matches.push({
              patternName: 'High-entropy string',
              rulePattern: 'high-entropy',
              matched: heMatch,
              line: lineIdx + 1,
              column: line.indexOf(heMatch) + 1,
            });
          }
        }
      }

      // Convert matches to findings
      for (const match of matches) {
        const matchedRules = rulesByPattern.get(match.rulePattern);
        if (!matchedRules || matchedRules.length === 0) continue;

        for (const rule of matchedRules) {
          findings.push({
            id: rule.id,
            title: rule.title,
            description: `${rule.description} (${match.patternName}: ${redact(match.matched)})`,
            severity: rule.severity,
            framework: rule.framework,
            controlIds: rule.controls,
            file: file.relativePath,
            line: match.line,
            column: match.column,
            evidence: redact(match.matched),
            analyzer: this.name,
            fixable: rule.fixable,
          });
        }
      }
    }

    return findings;
  }
}

File: products/compliance/analyzers/secret_scanner.go
package analyzers

import (
	"fmt"
	"math"
	"os"
	"regexp"
	"strings"

	"github.com/rishav1305/soul/products/compliance/rules"
)

// secretPattern links a compiled regex to a rule pattern identifier.
type secretPattern struct {
	name        string
	regex       *regexp.Regexp
	rulePattern string
}

// textExtensions lists file extensions that the scanner will inspect.
var textExtensions = map[string]bool{
	"ts":         true,
	"js":         true,
	"py":         true,
	"go":         true,
	"java":       true,
	"rb":         true,
	"yaml":       true,
	"yml":        true,
	"json":       true,
	"toml":       true,
	"env":        true,
	"cfg":        true,
	"conf":       true,
	"ini":        true,
	"xml":        true,
	"properties": true,
}

const maxFileSize int64 = 500 * 1024 // 500 KB

// secretPatterns contains all 16 regex patterns compiled at package init time.
var secretPatterns = []secretPattern{
	// 1. AWS Access Key
	{
		name:        "AWS Access Key",
		regex:       regexp.MustCompile(`AKIA[0-9A-Z]{16}`),
		rulePattern: "hardcoded-credential",
	},
	// 2. AWS Secret Key
	{
		name:        "AWS Secret Key",
		regex:       regexp.MustCompile(`(?:aws_secret|AWS_SECRET|secret_key|SECRET_KEY)\s*[=:]\s*['"]?([A-Za-z0-9/+=]{40})['"]?`),
		rulePattern: "hardcoded-credential",
	},
	// 3. GitHub Token
	{
		name:        "GitHub Token",
		regex:       regexp.MustCompile(`gh[ps]_[A-Za-z0-9_]{36,}`),
		rulePattern: "api-token",
	},
	// 4. Private Key
	{
		name:        "Private Key",
		regex:       regexp.MustCompile(`-----BEGIN\s+(?:RSA|EC|DSA|PGP)?\s*PRIVATE KEY-----`),
		rulePattern: "private-key",
	},
	// 5. JWT Token
	{
		name:        "JWT Token",
		regex:       regexp.MustCompile(`eyJ[A-Za-z0-9\-_]+\.eyJ[A-Za-z0-9\-_]+\.[A-Za-z0-9\-_.+/=]+`),
		rulePattern: "api-token",
	},
	// 6. Slack Token
	{
		name:        "Slack Token",
		regex:       regexp.MustCompile(`xox[bpras]-[0-9a-zA-Z-]+`),
		rulePattern: "api-token",
	},
	// 7. Stripe Key
	{
		name:        "Stripe Key",
		regex:       regexp.MustCompile(`sk_(?:live|test)_[0-9a-zA-Z]{24,}`),
		rulePattern: "api-token",
	},
	// 8. Anthropic Key
	{
		name:        "Anthropic Key",
		regex:       regexp.MustCompile(`sk-ant-[a-zA-Z0-9-]+`),
		rulePattern: "api-token",
	},
	// 9. Generic Password (case insensitive)
	{
		name:        "Generic Password",
		regex:       regexp.MustCompile(`(?i)(?:password|passwd|pwd|secret)\s*[=:]\s*['"][^'"]{4,}['"]`),
		rulePattern: "hardcoded-credential",
	},
	// 10. Generic API Key (case insensitive)
	{
		name:        "Generic API Key",
		regex:       regexp.MustCompile(`(?i)(?:api[_\-]?key|apikey)\s*[=:]\s*['"][^'"]{8,}['"]`),
		rulePattern: "hardcoded-credential",
	},
	// 11. Database URL
	{
		name:        "Database URL",
		regex:       regexp.MustCompile(`(?:mongodb|postgres|mysql|redis)://[^\s'"]+:[^\s'"]+@`),
		rulePattern: "hardcoded-credential",
	},
	// 12. Google API Key
	{
		name:        "Google API Key",
		regex:       regexp.MustCompile(`AIza[0-9A-Za-z\-_]{35}`),
		rulePattern: "api-token",
	},
	// 13. Heroku API Key (UUID)
	{
		name:        "Heroku API Key",
		regex:       regexp.MustCompile(`[0-9a-fA-F]{8}-[0-9a-fA-F]{4}-[0-9a-fA-F]{4}-[0-9a-fA-F]{4}-[0-9a-fA-F]{12}`),
		rulePattern: "api-token",
	},
	// 14. SendGrid API Key
	{
		name:        "SendGrid API Key",
		regex:       regexp.MustCompile(`SG\.[0-9A-Za-z\-_]{22,}\.[0-9A-Za-z\-_]{22,}`),
		rulePattern: "api-token",
	},
	// 15. Twilio API Key
	{
		name:        "Twilio API Key",
		regex:       regexp.MustCompile(`SK[0-9a-fA-F]{32}`),
		rulePattern: "api-token",
	},
	// 16. Mailgun API Key
	{
		name:        "Mailgun API Key",
		regex:       regexp.MustCompile(`key-[0-9a-zA-Z]{32}`),
		rulePattern: "api-token",
	},
}

// highEntropyRegex matches quoted strings of 20+ alphanumeric/base64 characters.
var highEntropyRegex = regexp.MustCompile(`['"]([A-Za-z0-9+/=\-_]{20,})['"]`)

// hexLikeRegex checks if a string is composed entirely of hex characters.
var hexLikeRegex = regexp.MustCompile(`^[0-9a-fA-F]+$`)

// base64LikeRegex checks if a string is composed entirely of base64 characters.
var base64LikeRegex = regexp.MustCompile(`^[A-Za-z0-9+/=\-_]+$`)

// matchResult holds a single pattern match within a file.
type matchResult struct {
	patternName string
	rulePattern string
	matched     string
	line        int
	column      int
}

// SecretScanner implements the Analyzer interface for detecting hardcoded
// secrets, credentials, API tokens, and high-entropy strings.
type SecretScanner struct{}

// Name returns the analyzer identifier used to match rules.
func (s *SecretScanner) Name() string {
	return "secret-scanner"
}

// Analyze scans the provided files for secrets using regex patterns and
// Shannon entropy detection, returning findings mapped to the relevant rules.
func (s *SecretScanner) Analyze(files []ScannedFile, allRules []rules.Rule) ([]Finding, error) {
	myRules := filterRules(allRules, s.Name())
	if len(myRules) == 0 {
		return nil, nil
	}

	rulesByPattern := groupRulesByPattern(myRules)
	var findings []Finding

	for _, file := range files {
		// Skip non-text files
		if !isTextExtension(file.Extension) {
			continue
		}
		// Skip large files
		if file.Size > maxFileSize {
			continue
		}

		content, err := os.ReadFile(file.Path)
		if err != nil {
			// Skip files that can't be read
			continue
		}

		// Quick binary check: if content contains null bytes, skip
		if containsNullByte(content) {
			continue
		}

		text := string(content)
		lines := strings.Split(text, "\n")
		var matches []matchResult

		for lineIdx, line := range lines {
			// Run each secret pattern against the line
			for _, pat := range secretPatterns {
				locs := pat.regex.FindAllStringIndex(line, -1)
				for _, loc := range locs {
					matches = append(matches, matchResult{
						patternName: pat.name,
						rulePattern: pat.rulePattern,
						matched:     line[loc[0]:loc[1]],
						line:        lineIdx + 1,
						column:      loc[0] + 1,
					})
				}
			}

			// Check for high-entropy strings
			heMatches := detectHighEntropyStrings(line)
			for _, heMatch := range heMatches {
				// Avoid duplicate findings: skip if already matched by a specific pattern
				alreadyMatched := false
				for _, m := range matches {
					if m.line == lineIdx+1 && strings.Contains(line, m.matched) && strings.Contains(heMatch, m.matched) {
						alreadyMatched = true
						break
					}
				}
				if !alreadyMatched {
					col := strings.Index(line, heMatch)
					matches = append(matches, matchResult{
						patternName: "High-entropy string",
						rulePattern: "high-entropy",
						matched:     heMatch,
						line:        lineIdx + 1,
						column:      col + 1,
					})
				}
			}
		}

		// Convert matches to findings
		for _, m := range matches {
			matchedRules, ok := rulesByPattern[m.rulePattern]
			if !ok || len(matchedRules) == 0 {
				continue
			}

			for _, rule := range matchedRules {
				findings = append(findings, Finding{
					ID:          rule.ID,
					Title:       rule.Title,
					Description: fmt.Sprintf("%s (%s: %s)", rule.Description, m.patternName, redact(m.matched)),
					Severity:    rule.Severity,
					Framework:   rule.Framework,
					ControlIDs:  rule.Controls,
					File:        file.RelativePath,
					Line:        m.line,
					Column:      m.column,
					Evidence:    redact(m.matched),
					Analyzer:    s.Name(),
					Fixable:     rule.Fixable,
				})
			}
		}
	}

	return findings, nil
}

// filterRules returns only the rules that belong to the named analyzer.
func filterRules(allRules []rules.Rule, analyzerName string) []rules.Rule {
	var filtered []rules.Rule
	for _, r := range allRules {
		if r.Analyzer == analyzerName {
			filtered = append(filtered, r)
		}
	}
	return filtered
}

// groupRulesByPattern builds a map from pattern string to the rules that use it.
func groupRulesByPattern(ruleList []rules.Rule) map[string][]rules.Rule {
	m := make(map[string][]rules.Rule)
	for _, r := range ruleList {
		m[r.Pattern] = append(m[r.Pattern], r)
	}
	return m
}

// redact keeps the first 4 and last 4 characters, replacing the middle with ****.
// If the string is 8 characters or fewer, it returns "****".
func redact(secret string) string {
	if len(secret) <= 8 {
		return "****"
	}
	return secret[:4] + "****" + secret[len(secret)-4:]
}

// isTextExtension checks whether the extension is one that should be scanned.
func isTextExtension(ext string) bool {
	return textExtensions[strings.ToLower(ext)]
}

// containsNullByte checks if the content contains any null bytes (binary indicator).
func containsNullByte(data []byte) bool {
	for _, b := range data {
		if b == 0 {
			return true
		}
	}
	return false
}

// shannonEntropy computes the Shannon entropy of a string.
func shannonEntropy(s string) float64 {
	if len(s) == 0 {
		return 0
	}
	freq := make(map[rune]int)
	for _, ch := range s {
		freq[ch]++
	}
	length := float64(len([]rune(s)))
	entropy := 0.0
	for _, count := range freq {
		p := float64(count) / length
		entropy -= p * math.Log2(p)
	}
	return entropy
}

// isHexLike checks if a string is composed entirely of hex characters.
func isHexLike(s string) bool {
	return hexLikeRegex.MatchString(s)
}

// isBase64Like checks if a string is composed entirely of base64 characters.
func isBase64Like(s string) bool {
	return base64LikeRegex.MatchString(s)
}

// detectHighEntropyStrings finds quoted strings with high Shannon entropy
// that may be secrets.
func detectHighEntropyStrings(line string) []string {
	var results []string
	submatches := highEntropyRegex.FindAllStringSubmatchIndex(line, -1)
	for _, loc := range submatches {
		// loc[2] and loc[3] are the start and end of capture group 1
		candidate := line[loc[2]:loc[3]]
		if len(candidate) < 20 {
			continue
		}

		var threshold float64
		if isHexLike(candidate) {
			threshold = 4.5
		} else if isBase64Like(candidate) {
			threshold = 5.0
		} else {
			threshold = 5.0
		}

		if shannonEntropy(candidate) > threshold {
			results = append(results, candidate)
		}
	}
	return results
}
